src/Symfony/Component/Semaphore/Store/LockStore.php

```diff
@@ -0,0 +1,134 @@
+            $lock->release();
+        }
+        $key->setState(__CLASS__, null);
+    }
+
+    public function exists(Key $key): bool
+    {
+        $locks = $this->getLocks($key, 0, create: false);
+
+        if (\count($locks) !== $key->getWeight()) {
+            foreach ($locks as $lock) {
+                $lock->release();
+            }
+
+            return false;
+        }
+
+        return true;
+    }
+
+    public function putOffExpiration(Key $key, float $ttlInSecond): void
+    {
+        $locks = $this->getLocks($key, $ttlInSecond, create: false);
+        foreach ($locks as $lock) {
+            if ($lock->isExpired()) {
+                foreach ($locks as $lock) {
+                    $lock->release();
+                }
+
+                throw new SemaphoreExpiredException($key, 'One or multiple locks are already expired.');
+            }
+
+            $lock->refresh($ttlInSecond);
+        }
+
+        if (\count($locks) !== $key->getWeight()) {
+            foreach ($locks as $lock) {
+                $lock->release();
+            }
+
+            throw new SemaphoreExpiredException($key, 'One or multiple locks were not even acquired.');
+        }
+    }
+
+    /**
+     * @return array<LockInterface>
+     */
+    private function getLocks(Key $key, float $ttlInSecond, bool $create): array
+    {
+        $lockName = base64_encode($key->__toString());
+
+        $locks = [];
+        if (!$key->hasState(__CLASS__) && $create) {
+            $limit = $key->getLimit();
+
+            // use a random start point to have a higher chance to catch a free slot directly
+            $startPoint = rand(0, $limit - 1);
+
+            for ($i = 0; $i < $limit; ++$i) {
+                $index = ($startPoint + $i) % $limit;
+
+                $lock = $this->lockFactory->createLock($lockName.'_'.$index, $ttlInSecond);
+                if ($lock->acquire(blocking: false)) { // use lock if we can acquire it else try to catch next lock
+                    $locks[] = $lock;
+
+                    if (\count($locks) === $key->getWeight()) {
+                        break;
+                    }
+                }
+            }
+
+            if (\count($locks) !== $key->getWeight()) {
+                foreach ($locks as $lock) {
+                    $lock->release();
+                }
+
+                throw new SemaphoreAcquiringException($key, 'There where no free locks found');
+            }
+
+            $key->setState(__CLASS__, $locks);
+        } elseif ($key->hasState(__CLASS__)) {
+            $locks = $key->getState(__CLASS__);
+            \assert((function () use ($locks) {
+                foreach ($locks as $lock) {
+                    if (!$lock instanceof LockInterface) {
+                        return false;
+                    }
+                }
+
+                return true;
+            })() === true, 'Locks must be an array of LockInterface');
```

Comment:
how is it possible to get something else in the state?

IIUC the state is set by this class at line 118, so it's always a LockInterface.

Reply:
Its just here for typesafety as return type is mixed, but can completly be removed.

Reply:
Removed it looks psalm still fine with it.